demos/text_detection_demo/cpp_gapi/src/text_recognition.cpp

```diff
@@ -0,0 +1,208 @@
+    return f.good();
+}
+bool tryReadVocabFile(const std::string& filename, std::string& alphabet) {
+    if (fileExists(filename)) {
+        alphabet = "";
+        std::ifstream inputFile(filename);
+        if (!inputFile.is_open()) {
+            throw std::runtime_error("Can't open the vocab file: " + filename);
+        }
+        std::string vocabLine;
+        while (std::getline(inputFile, vocabLine)) {
+            if (vocabLine.length() != 1) {
+                throw std::invalid_argument("Lines in the vocab file must contain 1 character");
+            }
+            alphabet += vocabLine;
+        }
+        if (alphabet.empty()) {
+            throw std::logic_error("File is empty: " + filename);
+        }
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void softmaxAndChoose(const std::vector<float>::const_iterator& begin,
+                      const std::vector<float>::const_iterator& end,
+                      int *argmax, float *prob) {
+    auto maxElem = std::max_element(begin, end);
+    *argmax = static_cast<int>(std::distance(begin, maxElem));
+    float maxVal = *maxElem;
+    double sum = 0;
+    for (auto i = begin; i != end; i++) {
+        sum += std::exp((*i) - maxVal);
+    }
+    if (std::fabs(sum) < std::numeric_limits<double>::epsilon()) {
+        throw std::logic_error("sum can't be equal to zero");
+    }
+    *prob = 1.0f / static_cast<float>(sum);
+}
+
+std::vector<float> softmax(const std::vector<float>::const_iterator& begin,
+                           const std::vector<float>::const_iterator& end) {
+    std::vector<float> prob(end - begin, 0.f);
+    std::transform(begin, end, prob.begin(), [](float x) { return std::exp(x); });
+    float sum = std::accumulate(prob.begin(), prob.end(), 0.0f);
+    for (int i = 0; i < static_cast<int>(prob.size()); i++)
+        prob[i] /= sum;
+    return prob;
+}
+
+struct BeamElement {
+    std::vector<int> sentence;  //!< The sequence of chars that will be a result of the beam element
+    float probBlank;            //!< The probability that the last char in CTC sequence
+                                //!< for the beam element is the special blank char
+    float probNotBlank;         //!< The probability that the last char in CTC sequence
+                                //!< for the beam element is NOT the special blank char
+
+    float prob() const {        //!< The probability of the beam element.
+        return probBlank + probNotBlank;
+    }
+};
+
+std::string SimpleDecoder(const std::vector<float>& data, const std::string& alphabet,
+                          char padSymbol, double *conf, int startIdx) {
+    std::string result = "";
+    const int numClasses = alphabet.length();
+    *conf = 1;
+
+    for (auto it = data.begin() + startIdx * numClasses; it != data.end(); it += numClasses) {
+        int argmax;
+        float prob;
+
+        softmaxAndChoose(it, it + numClasses, &argmax, &prob);
+        (*conf) *= prob;
```

Comment:
MB the reference will be simpler the pointer.

Reply:
OK